server/src/app.ts

```diff
@@ -66,6 +141,42 @@ app.get('/api/config', configController(config));
 app.get('/api/token', tokenController(identityClient, config));
 app.use('/api/rooms', roomsRouter(identityClient, roomsClient));
 
+// Function to send events to all connected clients
+export const sendEventToClients = (event: string, data: Record<string, unknown>): void => {
+  const message = `event: ${event}\ndata: ${JSON.stringify(data)}\n\n`;
+  console.log(`Sending ${event} event to ${clients.length} clients`);
+  // Copy array to avoid issues if array is modified during iteration
+  const clientsCopy = [...clients];
+  // Track successful sends for logging
+  let successCount = 0;
+  clientsCopy.forEach((client, index) => {
+    try {
+      // Check if client connection is still valid before writing
+      if (!client.writableEnded && !client.closed) {
+        // First flush any pending data to ensure event ordering
+        client.flushHeaders();
+        // Then write the new event
+        client.write(message);
+        successCount++;
+      } else {
+        console.log(`Client ${index} connection ended, removing from clients list`);
+        const clientIndex = clients.indexOf(client);
+        if (clientIndex !== -1) {
+          clients.splice(clientIndex, 1);
+        }
+      }
+    } catch (error) {
+      console.error(`Failed to send event to client ${index}:`, error);
+      // Remove failed client from the original array
```

Comment:
Is it 100% of any error we want to ignore that client? What if the client went into a tunnel for 2 mins, would it error then but reconnect?

Reply:
so the client will initiate a reconnect when it has a internet connection again. this is so the server isn't blasting bits at a dead connection. EventSource will attempt to reconnect to the subscription on the client side as part of the API